### PR TITLE
refactor to use bfe.json for sigil/suffix checks etc

### PR DESCRIPTION
**Pull request by mixmix**
This PR:
- changes how encoding happens - shifting it to use `bfe.json` as source of truth for sigil/ suffix checks
- creates custom regex matchers to ensure type/format matching with sigil/suffix strings
    - NOTE - this currently assumes data content of strings is base64 encoded (which is currently a safe assumption)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,284 +1,161 @@
...
+  if (input === null) return null
+  if (Number.isInteger(input)) return input
+
+  if (Buffer.isBuffer(input)) {
     if (input.length < 2)
       throw new Error(
         'Buffer is missing first two type&format fields: ' + input
       )
-    const t = input.slice(0, 1)
-    const tf = input.slice(0, 2)
```
Comment (staltz) [inline index.js:254]: Could you bring back this pattern? It was a nice once, reviewed by others in the team and merged, and your refactor of it seems unrelated to the core purpose of this PR. Repeating `input.slice(0, 2)` everywhere is error prone, not as clear as `tf` and not as performant.

File: index.js [hunk 2]
```diff
@@ -1,284 +1,161 @@
-// Following the TFD spec (also known as TFK), the naming convention in this
+// Following the BFE spec (formally known as TFK), the naming convention in this
```
Comment (staltz) [inline index.js:1]: I think you meant "formerly"?

File: index.js [hunk 3]
```diff
@@ -1,284 +1,161 @@
...
+const toTF = (type, format) =>
+  Buffer.from([NAMED_TYPES[type].code, NAMED_TYPES[type].formats[format].code])
+
+const STRING_TF = toTF('generic', 'UTF8 string')
+const NIL_TF = toTF('generic', 'nil')
+const BOOL = {
+  TF: toTF('generic', 'boolean'),
+  TRUE: Buffer.from([1]),
+  FALSE: Buffer.from([0]),
+}
```
Comment (staltz) [inline index.js:24]: Can you bring back BOOL_TF and BOOL_TRUE? Would be consistent with what we have already (e.g. STRING_TF) and this seems unrelated to the core of this PR

File: index.js [hunk 4]
```diff
@@ -1,284 +1,161 @@
...
-    let msgExtension
-    if (tf.equals(CLASSIC_MSG_TF)) msgExtension = '.sha256'
-    else if (tf.equals(GABBYGR_MSG_TF)) msgExtension = '.ggmsg-v1'
-    else if (tf.equals(BENDYBT_MSG_TF)) msgExtension = '.bbmsg-v1'
-    else throw new Error('Unknown msg: ' + buf)
-
-    const b64part = d.toString('base64')
-    return '%' + b64part + msgExtension
+  sigilSuffix(input, type, format) {
+    return [
```
Comment (staltz) [inline index.js:88]: Can you bring back `const d = input.slice(2)` which was a pattern used everywhere here?

File: index.js [hunk 5]
```diff
@@ -1,284 +1,161 @@
...
-  } else if (typeof input === 'object' && input !== null) {
+
+    if (input.equals(NIL_TF)) return null
+    if (input.slice(0, 2).equals(BOOL.TF)) return decoder.bool(input)
+    if (input.slice(0, 2).equals(STRING_TF)) return decoder.string(input)
+
+    const type = TYPES.find((type) => type.code.equals(input.slice(0, 1)))
+    if (type) {
+      const format = type.formats.find((format) =>
+        format.code.equals(input.slice(1, 2))
```
Comment (staltz) [inline index.js:125]: `input.slice(1, 2)` could be called `f`

File: index.js [hunk 6]
```diff
@@ -1,284 +1,161 @@
...
+
+  // FIXME: more checks, including floats!
+  throw new Error("don't know how to decode: " + input)
 }
 
 module.exports = {
   encode,
   decode,
-  bfeTypes: TYPES,
+  toString: decode, // alias
```
Comment (staltz) [inline index.js:158]: `toString` is highly misleading because the decoder can return booleans, null, and buffers. I suggest removing this.

**Review by staltz — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (mixmix) [reply index.js:254]: yep

File: index.js (hunk 2 above)
Comment (mixmix) [reply index.js:1]: embarrsing! I'm the english speaker

File: index.js (hunk 3 above)
Comment (mixmix) [reply index.js:24]: I don't like it as much but sure

File: index.js (hunk 4 above)
Comment (mixmix) [reply index.js:88]: done

File: index.js (hunk 6 above)
Comment (mixmix) [reply index.js:158]: quite right.
removed

File: index.js (hunk 5 above)
Comment (mixmix) [reply index.js:125]: done

**Comment by mixmix** [comment]
done @staltz

**Review by staltz — APPROVED** [review]
Looks good! Thanks for the patience.

There is one mysterious thing, which is not a blocker for merging, just wanted to point it out: I don't know how arbitrary Buffers are *decoded* in this new code, but apparently they are correctly decoded, according to the passing tests.

**Review by staltz — CHANGES_REQUESTED** [review]
So, I checked out the mysterious thing about decoding arbitrary buffers, and this PR has a problem. Maybe our tests weren't thorough enough in 06-generic.js, they mostly covered encoding.

If you write

```js
console.log(bfe.decode(encoded[9]))
```

(where `encoded[9]` is the passed-through buffer), the execution will error with

```
/home/staltz/oss/ssb-bfe/index.js:138
      } else throw new Error(`Unknown ${type.type} format`)
             ^

Error: Unknown diffie-hellman format
```

Which makes sense because the test buffer is 030201, and 03-beginning things are diffie-hellman.

**Comment by mixmix** [comment]
Arbitrary buffer type is not in here yet is it?

**Comment by staltz** [comment]
> Arbitrary buffer type is not in here yet is it?

No, but currently in master branch running `console.log(bfe.decode(encoded[9])` will return `<Buffer 03 02 01>`, not error.

**Comment by staltz** [comment]
As context for everyone: @mixmix and I spoke on a video call, and agreed that I could continue working on this PR.

I added 3 commits:

- [fix the testing of decoding arbitrary buffers](https://github.com/ssb-ngi-pointer/ssb-bfe/pull/22/commits/4d526909c899d3c2fe598f0415d0527aca827df8) addresses what I [found here](https://github.com/ssb-ngi-pointer/ssb-bfe/pull/21/files#r686822043) which made the tests fail (as expected)
- [encode/decode arbitrary buffers with TF = 0x0603](https://github.com/ssb-ngi-pointer/ssb-bfe/pull/22/commits/872182b1289b0320830523feddbd2edaf5109c78) which will supersede PR #20
- [standardize all error messages](https://github.com/ssb-ngi-pointer/ssb-bfe/pull/22/commits/e7bfc01069851d8183eba7ec113b4a6d6a00ba95) because I noticed some error messages were still incorrect

Requesting a review from both @arj03 and @mixmix

**Review by staltz — APPROVED** [review]

**Review by arj03 — APPROVED** [review]
Just some minor stylistic comments that can easily be done in another PR if needed. Functionally this is great and much easier to read (the encoder almost fits on a screen now which is really nice).

File: index.js [hunk 7]
```diff
@@ -1,284 +1,185 @@
...
+  }
 
-  boolean(bool) {
-    const d = bool ? BOOL_TRUE : BOOL_FALSE
-    return Buffer.concat([BOOL_TF, d])
-  },
-}
+  if (typeof input === 'boolean') return encoder.boolean(input)
+  if (input === null) return encoder.nil()
+  if (Buffer.isBuffer(input)) return encoder.bytes(input)
```
[the remark this answers is not in the document]
Comment (staltz) [reply index.js:70]: Oh yeah, I didn't realize this was just `if`s. I can change this

[the remark this answers is not in the document]
Comment (staltz) [reply index.js:70]: Oh yeah, I didn't realize this was just `if`s. I'll do this, but in a different PR. It turns out that if I consistently use `else if` everywhere, then this whole file looks a lot different.

File: index.js [hunk 8]
```diff
@@ -1,284 +1,185 @@
...
 // and "D" to mean "Data bytes".
 
-const { isFeedType, isMsgType, isBlobType } = require('ssb-ref')
-const { definitionsToDict } = require('./util')
-const TYPES = require('./bfe.json')
+const definitions = require('./bfe.json')
+const {
+  decorateBFE,
+  definitionsToDict,
+  findClassicTypeFormat,
```
[the remark this answers is not in the document]
Comment (staltz) [reply index.js]: How about `findTypeFormatForSigilSuffix`? (See updated PR)

[the remark this answers is not in the document]
Comment (arj03) [reply index.js]: That is consistent with the other function so 👍 from me

File: index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (mixmix) [reply index.js:70]: yeah that was me. I have a prediliction for early-returns, and find the alignment that linters often force on if-else make reading way more clunky.
but yeah agree they are probably safer
